src/online2bin/online2-net-nnet3-latgen-faster.cc

```diff
@@ -0,0 +1,398 @@
+// online2bin/online2-wav-nnet3-latgen-faster.cc
+
+// Copyright 2014  Johns Hopkins University (author: Daniel Povey)
+//           2016  Api.ai (Author: Ilya Platonov)
+//           2018  Polish-Japanese Academy of Information Technology (Author: Danijel Korzinek)
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "feat/wave-reader.h"
+#include "online2/online-nnet3-decoding.h"
+#include "online2/online-nnet2-feature-pipeline.h"
+#include "online2/onlinebin-util.h"
+#include "online2/online-timing.h"
+#include "online2/online-endpoint.h"
+#include "fstext/fstext-lib.h"
+#include "lat/lattice-functions.h"
+#include "util/kaldi-thread.h"
+#include "nnet3/nnet-utils.h"
+
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <arpa/inet.h>
+#include <unistd.h>
+
+namespace kaldi {
+
+class TcpServer {
+ public:
+  TcpServer();
+  ~TcpServer();
+
+  bool Listen(int32 port);  //start listening on a given port
+  int32 Accept();  //accept a client and return its descriptor
+
+  bool ReadChunk(size_t len); //get more data and return false if end-of-stream
+  Vector<BaseFloat> getChunk(); //get the data read by above method
+
+  bool Write(std::string msg); //write to accepted client
+
+  void disconnect();
+
+ private:
+  struct ::sockaddr_in h_addr_;
+  int32 server_desc_, client_desc_;
+  int16 *samp_buf_;
+  size_t buf_len_, has_read_;
+};
+
+std::string LatticeToString(const Lattice &lat, fst::SymbolTable *word_syms) {
+  LatticeWeight weight;
+  std::vector<int32> alignment;
+  std::vector<int32> words;
+  GetLinearSymbolSequence(lat, &alignment, &words, &weight);
+
+  std::string msg = "";
+  for (size_t i = 0; i < words.size(); i++) {
+    msg += word_syms->Find(words[i]) + " ";
```

Comment:
Probably better to handle the case where `words[i]` is not found (e.g. show some error and return something meaningful to the client).

Reply:
Ok, will do.

Reply:
Fixed this.